src/selectors_vm/compiler.rs

```diff
@@ -23,46 +31,134 @@ pub struct CompiledAttributeExprOperand {
 macro_rules! curry_compile_expr_macro {
     ($negation:ident) => {
         macro_rules! compile_expr {
-            (|$arg:ident| $body:expr) => {
+            (|$state:tt, $arg:tt| $body:expr) => {
                 if $negation {
-                    Box::new(move |$arg| !($body))
+                    Box::new(move |$state, $arg| !($body))
                 } else {
-                    Box::new(move |$arg| $body)
+                    Box::new(move |$state, $arg| $body)
                 }
             };
 
             (@unmatchable) => {
-                compile_expr!(|_arg| false);
+                compile_expr!(|_, _| false)
             };
 
-            (@match_all) => {
-                compile_expr!(|_arg| true);
-            };
+            (@match_any) => {
+                compile_expr!(|_, _| true)
+            }
         }
-    };
+    }
 }
 
-trait CompileOr<T> {
-    fn compile_or(
-        self,
-        negation: bool,
-        compile: impl Fn(T) -> CompiledAttributeExpr,
-    ) -> CompiledAttributeExpr;
+trait Compilable {
+    fn compile(&self, encoding: &'static Encoding, exprs: &mut ExprSet, flags: &mut ProgramFlags);
 }
 
-impl<T> CompileOr<T> for Result<T, ()> {
-    #[inline]
-    fn compile_or(
-        self,
-        negation: bool,
-        compile: impl Fn(T) -> CompiledAttributeExpr,
-    ) -> CompiledAttributeExpr {
+impl Compilable for Expr<OnTagNameExpr> {
+    fn compile(
+        &self,
+        encoding: &'static Encoding,
+        ExprSet { local_name_exprs, .. }: &mut ExprSet,
+        flags: &mut ProgramFlags,
+    ) {
+        let &Self { negation, ref simple_expr } = self;
         curry_compile_expr_macro!(negation);
 
-        match self {
-            Ok(v) => compile(v),
-            Err(_) => compile_expr!(@unmatchable),
+        local_name_exprs.push(
+            match simple_expr {
```

Comment:
Avoid deep nesting of expressions: it significantly hurts readability of the code. Here result of the match arm can be assigned to a variable which later pushed to exprs vector. This allow to get rid of one nesting level.